specs/0000-table-hide-index-header/product-spec.md

```diff
@@ -0,0 +1,121 @@
+---
+author: "@lukasmasuch"
+created: 2025-12-02
+status: Draft
+---
+
+# Allow hiding index and header columns in `st.table`
+
+## Summary
+
+Add `hide_index: bool | None = None` and `hide_header: bool = False` parameters to `st.table`,
+enabling users to control the visibility of the index column and column headers.
+
+## Problem
+
+`st.table` and `st.dataframe` have historically been undifferentiated. We've been working to
+clarify their roles:
+
+- **`st.dataframe`**: Large datasets with interactive exploration (sorting, filtering, selection)
+- **`st.table`**: Small, text-based customizable tables for displaying a few metrics or key-value pairs
+
+To make `st.table` better for its intended use case, users need more control over visual
+clutter. Currently, `st.table` always shows the index column (even meaningless 0, 1, 2...
+indices) and column headers, which is problematic for:
+
+1. Compact metric displays where numeric indices add no value
+2. Key-value pair tables where headers are unnecessary
+3. Displaying single rows where column context is already clear
+
+**Requests:**
+
+- [#8235](https://github.com/streamlit/streamlit/issues/8235) - `hide_header` (32+ upvotes)
+- [#9251](https://github.com/streamlit/streamlit/issues/9251) - `hide_index` for `st.table` (14+ upvotes)
+- [#13185](https://github.com/streamlit/streamlit/issues/13185) - Add a description list element for viewing key-value data
+
+**Consistency gap:**
+
+`st.dataframe` and `st.data_editor` already support `hide_index`, but `st.table` does not.
+
+## Proposal
+
+### API
+
+```python
+st.table(
+    ...,
+    hide_index: bool | None = None,  # NEW
+    hide_header: bool = False,       # NEW
+)
+```
+
+### Parameters
+
+| Parameter | Type | Default | Description |
+|-----------|------|---------|-------------|
+| `hide_index` | `bool \| None` | `None` | Whether to hide the index column. If `None`, auto-hide default RangeIndex. |
+| `hide_header` | `bool` | `False` | Whether to hide the column headers row. |
+
+### Behavior
+
+**`hide_index`:**
+
+- `None` (default): Hide RangeIndex (0, 1, 2...), show custom indices
+- `True`: Always hide
+- `False`: Always show
+
+**`hide_header`:**
+
+- `False` (default): Show column headers
+- `True`: Hide all header rows, including all levels of the MultiIndex headers.
+
+### Examples
+
+**Auto-hide default index:**
+
+```python
+import pandas as pd
+import streamlit as st
+
+# RangeIndex is auto-hidden
+df = pd.DataFrame({"A": [1, 2], "B": [3, 4]})
+st.table(df)  # Index hidden
+
+# Custom index is shown
+df_custom = df.set_index(pd.Index(["row1", "row2"]))
+st.table(df_custom)  # Index shown
+```
+
+**Key-value display (no headers):**
+
+```python
+# Clean key-value table for metrics
+st.table({
+    "Price": "$145.00",
+    "Customer": "Bobby Jones",
+    "Address": "129 Market St, NYC",
+    "Store": "Trader Joe's"
+}, hide_header=True)
```

Comment:
In this case (or in similar cases e.g. if you pass a list of lists), can we automatically hide the header? I would love it if the header is hidden in most cases for `st.table` and only shows up if you really explicitly set it.

Reply:
(If there's a good way to do this, should probably set `hide_header=None` by default).